File: 0136_Single_Number/sol.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;

class Solution {
public:
    // Strategy: XOR
    // Time:     O(N)
    // Space:    O(N)

    int solve(vector<int>& nums) {
        // Strategy #2 {XOR}
        int result = 0;
        for(auto x : nums){
            result ^= x;
        } 
        return result;

        // Strategy #1
        // set <int> temp = {};
        // for (int i = 0; i < nums.size(); i++){
        //     if(temp.find(nums[i]) != temp.end()){
        //         temp.erase(nums[i]);
        //     }else{
        //         temp.insert(nums[i]);
        //     }
        // }
        // return *temp.begin();
    }
};

struct TestCase {
    vector<int> input1;
    int expected;
};

void run_tests() {
    Solution sol;
    vector<TestCase> testCases{
        {{2,2,1}, 1}, 
        {{4,1,2,1,2}, 4},
        {{1}, 1}
    };

    for(size_t i = 0; i < testCases.size(); ++i) {
        int result = sol.solve(testCases[i].input1);
        bool isCorrect = (result == testCases[i].expected);
        cout << (isCorrect ? "✅" : "❌") << " Test #" << i;
        cout << " | Expected: " << testCases[i].expected 
             << " | Result: " << result << endl;
    }
}

int main() {
    run_tests();
    return 0;
}

File: 0136_Single_Number/sol.go
package main

import (
	"fmt"
	"reflect"
)

/**
 * Strategy: XOR
 * Time:     O(N)
 * Space:    O(N)
 */
// EDIT HERE: Change function signature and return type as needed
func solve(nums []int) int {
	result := 0
	for _, x := range nums {
		result ^= x
	}
	return result
}

func main() {
	type testCase struct {
		input1 []int
		expect int
	}
	tests := []testCase{
		{input1: []int{2, 2, 1}, expect: 1},
		{input1: []int{4, 1, 2, 1, 2}, expect: 4},
		{input1: []int{1}, expect: 1},
	}

	for i, tc := range tests {
		// EDIT HERE: Update arguments to match your function signature
		got := solve(tc.input1)
		if !reflect.DeepEqual(got, tc.expect) {
			fmt.Printf("❌ Test %d Failed | Input: %v | Got: %v, Want: %v\n",
				i, tc.input1, got, tc.expect)
		} else {
			fmt.Printf("✅ Test %d Passed | Input: %v | Got: %v, Want: %v\n",
				i, tc.input1, got, tc.expect)
		}
	}
}
